util/fork_func.c

```diff
@@ -0,0 +1,75 @@
+#include <sys/wait.h>
+
+#include <stdint.h>
+#include <unistd.h>
+
+#include "warnp.h"
+
+#include "fork_func.h"
+
+/**
+ * fork_func(func, cookie):
+ * Fork and run ${func} in a new process, with ${cookie} as the sole argument.
+ */
+pid_t
+fork_func(int (* func)(void *), void * cookie)
+{
+	pid_t pid;
+
+	/* Fork */
+	switch (pid = fork()) {
+	case -1:
+		warnp("fork");
+		goto err0;
+	case 0:
+		/* In child process: Run the provided function, then exit. */
+		_exit(func(cookie));
+	default:
+		/* In parent process: do nothing else. */
+		break;
+	}
+
+	/* Success! */
+	return (pid);
+
+err0:
+	/* Failure! */
+	return (-1);
+}
+
+/**
+ * fork_func_wait(pid):
+ * Wait for the process ${pid} to finish.  Print any error arising from ${pid}.
+ * If ${pid} exited cleanly, return its exit code; otherwise, return -1.
+ */
+int
+fork_func_wait(pid_t pid)
+{
+	int status;
+	int rc = -1;
+
+	/* Wait for the process to finish. */
+	if (waitpid(pid, &status, 0) == -1) {
+		warnp("waitpid");
+		goto err0;
+	}
+
+	/* Print the error status, if applicable. */
+	if (WIFEXITED(status)) {
+		/* Child ${pid} exited cleanly. */
+		rc = WEXITSTATUS(status);
+	} else {
+		/* Child ${pid} did not exit cleanly. */
+
+		/* Warn about the reason for the unclean exit. */
+		if (WIFSIGNALED(status))
+			warn0("pid %jd: signalled with %i",
```

Comment:
Hold-over from when I was a TA: "use %i for integer, and %f for float".  (Because to a first-year engineering student, "decimal" means something like "1.23".)

Ok, I'll switch to `%d`.